clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/evaluation/TriggersMatcher.kt

```diff
@@ -0,0 +1,404 @@
+     *
+     * This function evaluates a single trigger condition by comparing the expected value with the actual value.
+     * The specified operator (`op`) is used to perform the comparison.
+     *
+     * @param op The [TriggerOperator] used for comparison.
+     * @param expected The expected [TriggerValue] for the condition.
+     * @param actual The actual [TriggerValue] to be compared with the expected value.
+     * @return `true` if the condition is satisfied, `false` otherwise.
+     */
+    @VisibleForTesting
+    internal fun evaluate(
+        op: TriggerOperator,
+        expected: TriggerValue,
+        actual: TriggerValue
+    ): Boolean {
+        if (actual.value == null) {
+            return op == TriggerOperator.NotSet
+        }
+
+        return when (op) {
+            TriggerOperator.Set -> true
+            TriggerOperator.LessThan -> expectedValueLessThanGreaterThanActual(expected, actual, true)
+            TriggerOperator.GreaterThan -> expectedValueLessThanGreaterThanActual(expected, actual, false)
+            TriggerOperator.Equals -> expectedValueEqualsActual(expected, actual)
+            TriggerOperator.NotEquals -> !expectedValueEqualsActual(expected, actual)
+            TriggerOperator.Between -> actualIsInRangeOfExpected(expected, actual)
+            TriggerOperator.Contains -> actualContainsExpected(expected, actual)
+            TriggerOperator.NotContains -> !actualContainsExpected(expected, actual)
+            else -> false // TODO: Implement all cases as per the backend evaluation and remove this line
+        }
+    }
+
+    /**
+     * Internal function to evaluate a haversine distance condition.
+     *
+     * This function evaluates if the haversine distance between two locations is within a specified radius.
+     * The haversine formula is used to compute the distance between two locations.
+     *
+     * @param radius The radius to check against, in kilometers.
+     * @param expected The expected location.
+     * @param actual The actual location.
+     * @return `true` if the haversine distance is within the specified radius, `false` otherwise.
+     */
+    @VisibleForTesting
+    internal fun evaluateDistance(radius: Double, expected: Location, actual: Location): Boolean {
+        val distance = Utils.haversineDistance(expected, actual)
+        return distance <= radius
+    }
+
+    /**
+     * Internal function to check if the expected value equals the actual value.
+     *
+     * This function checks if the expected value is equal to the actual value. It handles various
+     * data types including strings, numbers, and lists.
+     *
+     * @param expected The expected [TriggerValue].
+     * @param actual The actual [TriggerValue].
+     * @return `true` if the expected value equals the actual value, `false` otherwise.
+     */
+    @VisibleForTesting
+    internal fun expectedValueEqualsActual(expected: TriggerValue, actual: TriggerValue): Boolean {
+        return when {
```

Comment:
we should have exhaustive when in kotlin, lets define type in `TriggerValue` to have a sealed val type.